feat(manager/gradle): Add verification metadata update support to gradle manager

udakak opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->

Adds functionality to update verification metadata to the gradle manager. This adds the `--write-verification-metadata` argument in the updateArtifacts step, so gradle will generate the data as suggested [here](https://docs.gradle.org/current/userguide/dependency_verification.html#sec:bootstrapping-verification).

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

Closes #13575 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

udakak committed:
add gradle verification metadata support

udakak committed:
add tests

udakak committed:
add pgp to hashTypes

udakak committed:
add documentation

HonkingGoose commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -148,13 +154,26 @@ export async function updateArtifacts({
+          hashTypes.push(hashType);
+        }
+      });
+      cmd += ` --write-verification-metadata ${hashTypes.join(',')}`;
```
Quote from [Gradle Docs, Bootstrapping dependency verification](https://docs.gradle.org/current/userguide/dependency_verification.html#sec:bootstrapping-verification):
 
> Bootstrapping can either be used to create a file from the beginning, or also to _update_ an existing file with new information. Therefore, it’s recommended to always use the same parameters once you started bootstrapping.
> 
> The dependency verification file can be generated with the following CLI instructions:
> 
> `gradle --write-verification-metadata sha256 help`

How do we make sure we're always using the same command? It sounds like mixing different commands will cause problems.

Also it looks like Gradle recommends starting with the `help` command first.

> There are dependencies that Gradle _cannot_ discover this way. In particular, you will notice that the CLI above uses the `help` task. If you don’t specify any task, Gradle will automatically run the default task and generate a configuration file at the end of the build too.
>
> The difference is that Gradle _may_ discover more dependencies and artifacts depending on the tasks you execute. As a matter of fact, Gradle cannot automatically discover _detached configurations_,  which are basically dependency graphs resolved as an internal  implementation detail of the execution of a task: they are not, in  particular, declared as an input of the task because they effectively  depend on the configuration of the task at execution time.
>
> A good way to start is just to use the simplest task, `help`,  which will discover as much as possible, and if subsequent builds fail  with a verification error, you can re-execute generation with the  appropriate tasks to "discover" more dependencies.
>
> Gradle won’t verify either checksums or signatures of plugins which use their own HTTP clients. Only plugins which use the infrastructure provided by Gradle for performing requests will see their requests verified.

udakak replied:
I thought the `parameters` in the docs referred to the hash type, and that it would be fine to use the `dependencies` task that is being run anyway.

But your comment got me thinking that it would be safest to just have an extra exec call just for the verification metadata, separated from the lockfile command. That way we can be sure that the same task (`help`) is used every time.

viceice replied:
sounds good

udakak commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -148,13 +154,26 @@ export async function updateArtifacts({
+      const verificationMetadata = await readLocalFile(
+        verificationMetadataFile
+      );
+      ['md5', 'sha1', 'sha256', 'sha512', 'pgp'].forEach((hashType) => {
```
If someone sets up their verification metadata using unsafe checksums, I would expect Renovate to update them. But if you view this differently I will follow what you say, since I have no need for the outdated checksums myself.

Simply not updating them will lead to verification errors after dependency updates, though. How would you warn users about what is happening?

viceice replied:
just use sha256 is there is a weak hash algorithm was used.

Churro replied:
If you update existing entries that have both `sha1` and `sha256` with `./gradlew --console=plain -q --write-verification-metadata sha256 help` would that ignore the existing `sha1` or delete it?

udakak replied:
It won't delete anything, it won't even delete the sha256 if it changes for some reason. It will leave the (potentially wrong) sha1 and add the new sha256 value as `<also-trust>` element.

In my opinion this is fine, since this is just how gradle does this. The hash for an existing artifact shouldn't change and if it does, it should be checked manually anyway.

Churro replied:
Agreed. From a user perspective, it's counter-intuitive that renovate ignores `md5` or `sha1` digests but apart from that, the whole verification metadata functionality makes no sense if used with hash functions that are not collision resistant.

udakak committed:
Apply suggestions from code review

Co-authored-by: HonkingGoose <34918129+HonkingGoose@users.noreply.github.com>

viceice commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -96,7 +99,10 @@ export async function updateArtifacts({
   const lockFiles = fileList.filter((file) => isLockFile(file));
-  if (!lockFiles.length) {
+  const verificationMetadataFile = fileList.find((fileName) =>
+    fileName.endsWith('verification-metadata.xml')
```
```suggestion
    fileName.endsWith('gradle/verification-metadata.xml')
```
to not find any file which ends with same name

viceice commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -148,13 +154,26 @@ export async function updateArtifacts({
+      const verificationMetadata = await readLocalFile(
+        verificationMetadataFile
+      );
+      ['md5', 'sha1', 'sha256', 'sha512', 'pgp'].forEach((hashType) => {
```
use for-of loop

udakak committed:
call write metadata in separate command

udakak committed:
don't use unsafe hash types

udakak commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -36,7 +36,10 @@ async function getUpdatedLockfiles(
-    if (isLockFile(modifiedFile)) {
+    if (
+      isLockFile(modifiedFile) ||
+      modifiedFile.endsWith('verification-metadata.xml')
```
```suggestion
      modifiedFile.endsWith('gradle/verification-metadata.xml')
```

udakak committed:
use full path for finding metadata file

Co-authored-by: Michael Kriese <michael.kriese@visualon.de>

viceice commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -142,28 +148,52 @@ export async function updateArtifacts({
+        hashTypes.push('sha256');
+      }
+      const verificationMetadataCmd =
+        baseCmd + ` --write-verification-metadata ${hashTypes.join(',')} help`;
```
string template

udakak committed:
write verification metadata after lockfile

udakak commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -142,28 +148,52 @@ export async function updateArtifacts({
 
-    await writeLocalFile(packageFileName, newPackageFileContent);
-    await exec(cmd, { ...execOptions, ignoreStdout: true });
+    await exec(lockfileCmd, { ...execOptions, ignoreStdout: true });
```
Good point, makes more sense the other way around, I switched it.

udakak committed:
fix tests

Churro reviewed (commented):
I tested it on the [reproduction repo](https://github.com/dbast/gradle-minimal) linked in https://github.com/renovatebot/renovate/issues/13575. It updated `org.junit.jupiter:junit-jupiter` from 5.7.2 -> 5.10.0 and regenerated `verification-metadata.xml`.

The only issue I faced was that  instead of updating `<component group="org.junit.jupiter" name="junit-jupiter" version="5.7.2">` it added yet another `<component group="org.junit.jupiter" name="junit-jupiter" version="5.10.0">` in addition (+ plus further deps that `org.junit.jupiter:junit-jupiter:5.10.0` relies on). It didn't take out the existing entries related to 5.7.2.

I suppose this is because of https://github.com/gradle/gradle/issues/19228 but it makes me think whether it would make sense to also add `--clean` to the `--write-verification-metadata` command in order to force-run a clean build. It's not entirely obvious to me though if this is needed here or not (esp. when thinking about the lockfile update happening before). WDYT?

udakak commented:
> I tested it on the [reproduction repo](https://github.com/dbast/gradle-minimal) linked in #13575. It updated `org.junit.jupiter:junit-jupiter` from 5.7.2 -> 5.10.0 and regenerated `verification-metadata.xml`.
> 
> The only issue I faced was that instead of updating `<component group="org.junit.jupiter" name="junit-jupiter" version="5.7.2">` it added yet another `<component group="org.junit.jupiter" name="junit-jupiter" version="5.10.0">` in addition (+ plus further deps that `org.junit.jupiter:junit-jupiter:5.10.0` relies on). It didn't take out the existing entries related to 5.7.2.
> 
> I suppose this is because of [gradle/gradle#19228](https://github.com/gradle/gradle/issues/19228) but it makes me think whether it would make sense to also add `--clean` to the `--write-verification-metadata` command in order to force-run a clean build. It's not entirely obvious to me though if this is needed here or not (esp. when thinking about the lockfile update happening before). WDYT?

This is actually how gradle does this, it never removes entries, only adds. To get rid of old entries we would need to delete the file and recreate it anew every time, which is probably not what we want. (https://docs.gradle.org/current/userguide/dependency_verification.html#sec:verification-metadata-hygiene)

Your linked issue refers to creating the file for the first time (if I did not misread), but it could also be a problem whenever dependencies are first added and then removed again, so that they are still in the cache (would this ever happen during a renovate run?)

udakak committed:
Apply suggestions from code review

Co-authored-by: Johannes Feichtner <343448+Churro@users.noreply.github.com>

udakak committed:
extract verification update method

udakak committed:
add test for default hashType

viceice commented:
> > I tested it on the [reproduction repo](https://github.com/dbast/gradle-minimal) linked in #13575. It updated `org.junit.jupiter:junit-jupiter` from 5.7.2 -> 5.10.0 and regenerated `verification-metadata.xml`.
> > 
> > The only issue I faced was that instead of updating `<component group="org.junit.jupiter" name="junit-jupiter" version="5.7.2">` it added yet another `<component group="org.junit.jupiter" name="junit-jupiter" version="5.10.0">` in addition (+ plus further deps that `org.junit.jupiter:junit-jupiter:5.10.0` relies on). It didn't take out the existing entries related to 5.7.2.
> > 
> > I suppose this is because of [gradle/gradle#19228](https://github.com/gradle/gradle/issues/19228) but it makes me think whether it would make sense to also add `--clean` to the `--write-verification-metadata` command in order to force-run a clean build. It's not entirely obvious to me though if this is needed here or not (esp. when thinking about the lockfile update happening before). WDYT?
> 
> This is actually how gradle does this, it never removes entries, only adds. To get rid of old entries we would need to delete the file and recreate it anew every time, which is probably not what we want. (https://docs.gradle.org/current/userguide/dependency_verification.html#sec:verification-metadata-hygiene)
> 
> Your linked issue refers to creating the file for the first time (if I did not misread), but it could also be a problem whenever dependencies are first added and then removed again, so that they are still in the cache (would this ever happen during a renovate run?)

this is a good candidate for lockfile maintenance, there that file should be deleted before writing it again.

viceice commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -153,17 +187,23 @@ export async function updateArtifacts({
 
     await writeLocalFile(packageFileName, newPackageFileContent);
-    await exec(cmd, { ...execOptions, ignoreStdout: true });
+    await exec(lockfileCmd, { ...execOptions, ignoreStdout: true });
+
+    await updateVerificationMetadata(
+      verificationMetadataFile,
+      baseCmd,
+      execOptions
+    );
```
You should combine both commands to a single exec call, you can pass an array of commands.
This will improve performance when on docker mode (eg: hosted app).
Currently this will cause two docker sidecar container, which both need to install tools.

viceice commented:
btw: use the re-request review function from github 😉
![image](https://github.com/renovatebot/renovate/assets/1798109/0ee46014-af72-4690-abde-b700467a014d)
Otherwise we normally don't review again without a request

udakak committed:
use pgp when signature verification is enabled

all keys could be grouped so that no <pgp> element exists,
so we need to check for <verify-signatures> instead

udakak committed:
call lockfile and verification cmd in one exec

udakak commented:
> this is a good candidate for lockfile maintenance, there that file should be deleted before writing it again.

I don't think we should ever delete this file automatically. It is not a lockfile, but a security feature. Knowing more hashes doesn't harm (except for wasted storage space), but deleting the wrong ones can.

Some dependencies might deliver different artifacts for different OSs or CPU architectures, while gradle will only be able to discover the artifacts for the current system. (So a dev might need an extra hash for the mac artifact, while the CI pipeline needs the linux version, we had problems with this in the past)
Some dependencies might not be discovered correctly by `gradle help`.
Maybe someone added something manually (like a trusted key for internal packages).

So automatically deleting the file and discarding any manual changes, will break the build for some users.

HonkingGoose commented on lib/modules/manager/gradle/readme.md:
```diff
@@ -8,3 +8,8 @@ During [lock file maintenance](https://docs.renovatebot.com/configuration-option
+### Dependency verification
+
+If Renovate finds a `gradle/verification-metadata.xml` file, it updates the content by using the `gradle --write-verification-metadata <hashTypes>` command.
+Renovate will check the file for existing hash types (like `sha256`) and use them as `<hashTypes>`.
```
Looks like we only check `sha256`, `sha512` and `pgp` style hashes?

We should explain _which_ hash types Renovate can check and update. Or users will ask why their `md5` or `sha1` hash is not found/updated. 😄 

```ts
async function buildUpdateVerificationMetadataCmd(
  verificationMetadataFile: string | undefined,
  baseCmd: string
): Promise<string | undefined> {
  if (verificationMetadataFile) {
    const hashTypes: string[] = [];
    const verificationMetadata = await readLocalFile(verificationMetadataFile);
    for (const hashType of ['sha256', 'sha512']) {
      if (verificationMetadata?.includes(`<${hashType}`)) {
        hashTypes.push(hashType);
      }
    }
    if (!hashTypes.length) {
      hashTypes.push('sha256');
    }
    if (
      verificationMetadata?.includes(
        '<verify-signatures>true</verify-signatures>'
      )
    ) {
      hashTypes.push('pgp');
    }
    return `${baseCmd} --write-verification-metadata ${hashTypes.join(
      ','
    )} help`;
  }
}
```

Churro replied:
I'd suggest adding a note that hash functions that gradle also supports verification metadata with `md5` and `sha1` but renovate doesn't take those into account since they are prone to collision attacks. In case hashes are encountered that were generated with any of these algorithms, renovate defaults to `sha256`.

HonkingGoose replied:
@Churro Good idea! Please review this suggestion:

```suggestion
Renovate will check the file for existing hash types (like `sha256`) and use them as `<hashTypes>`.

<!-- prettier-ignore -->
!!! warning
    Gradle allows verification metadata to use the `md5` and `sha1` algorithms.
    Because those algorithms are prone to collision attacks, Renovate ignores them.
    If Renovate encounters hashes that are generated with `md5` or `sha1` algorithms, Renovate uses `sha256` instead.
```

Churro commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -153,17 +189,27 @@ export async function updateArtifacts({
     await writeLocalFile(packageFileName, newPackageFileContent);
-    await exec(cmd, { ...execOptions, ignoreStdout: true });
+
+    const cmds = [lockfileCmd];
```
Although it makes sense to be used together, gradle dependency verification also works when gradle lockfiles are not used. In that case, parsing the list of subprojects and calling gradle to update a lockfile are obsolete. Running the commands anyway does no harm besides prolonging renovate runs.

I think it would make sense to run lockfile-related commands only when lockfiles are present. Could be a follow-up refactoring PR but if someone else thinks this should be done here, that's fine for me too.

viceice replied:
sounds good

Churro commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -86,6 +89,34 @@ async function getGradleVersion(gradlewFile: string): Promise<string | null> {
+    if (!hashTypes.length) {
+      hashTypes.push('sha256');
+    }
+    if (
```
Since metadata verification can also be toggled, wouldn't it make sense to also do a similar check for `<verify-metadata>true</verify-metadata>`? Adding hash algorithms only makes sense if this condition is fulfilled.

viceice replied:
good idea

udakak committed:
only call necessary commands

udakak committed:
add warning about ignored hash types

Co-authored-by: HonkingGoose <34918129+HonkingGoose@users.noreply.github.com>

Churro commented on lib/modules/manager/gradle/artifacts.spec.ts:
```diff
@@ -545,4 +548,239 @@ describe('modules/manager/gradle/artifacts', () => {
     ]);
   });
+
+  it('updates verification metadata file', async () => {
```
_Nitpick:_ would you please collect the dependency verification related tests in a dedicated component? 

Somewhat like this:
```ts
describe('dependency verification tests', () => {
  it('updates verification metadata file', async () => ...
  it('updates verification metadata and lock file', async () => ...
}
```` 

Since `artifact.ts` now implements a second use-case, it would help distinguishing the tests for verification metadata updates.

Churro commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -142,28 +192,49 @@ export async function updateArtifacts({
 
-      cmd += ` --update-locks ${updatedDepNames.map(quote).join(',')}`;
+    if (!cmds.length) {
+      logger.debug('No lockfile update necessary');
```
```suggestion
      logger.debug('No lockfile or verification metadata update necessary');
```

Churro commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -142,28 +192,49 @@ export async function updateArtifacts({
+    let lockfileCmd = '';
+    if (lockFiles.length) {
+      const subprojects = await getSubProjectList(baseCmd, execOptions);
+      lockfileCmd = `${baseCmd} ${subprojects
```
See previous suggested change

```suggestion
      let lockfileCmd = `${baseCmd} ${subprojects
```

Churro commented on lib/modules/manager/gradle/artifacts.ts:
```diff
@@ -142,28 +192,49 @@ export async function updateArtifacts({
-      .map(quote)
-      .join(' ')}`;
+    const cmds = [];
+    let lockfileCmd = '';
```
Only used within `if (lockFiles.length) { ... }`

```suggestion

```

Churro reviewed (commented):
Only a few minor things, otherwise LGTM

udakak committed:
Apply suggestions from code review

Co-authored-by: Johannes Feichtner <343448+Churro@users.noreply.github.com>

udakak committed:
distinguish between lockfile and verification tests

udakak committed:
Merge branch 'main' into feature/gradle-verification-metadata
